[dnf5] Repo code cleanups and separating PackageDownloader

lukash opened the pull request.

Commit:
libdnf/repo: Simplify calling download_metadata()

download_metadata() was setting the LRO_YUMDLIST option LR_RPMMD_FULL,
which is actually the default (LR_RPMMD_FULL is defined as NULL and the
yumdlist variable, which holds the value, is initialized to NULL via a
lr_malloc0() call in lr_handle_init()).

Therefore we can use a single method for downloading the metadata and
it doesn't need the LrHandle argument.

Commit:
libdnf/advisory: Don't use glib's g_strcmp0()

The char * is checked for nullptr already. Also adds braces.

Commit:
libdnf/repo: Move package downloading code into the PackageDownloader class

Commit:
utils: Clean up The TempDir class

Drop support for creating subdirectories, those are easy to create
manually, are only needed in tests and are beyong the functionality of
creating a temporary directory.

Also deletes the copy assignment operator and adds stopgap error
handling.

Commit:
Use TempDir in Repo downloading code

Adds a destdir argument to the TempDir class and uses the class for
temporary directories in Repo downloading code.

Commit:
libdnf/repo: Share code for creating LrHandle

The new_handle() function duplicated most of the code used also in
Repo::Impl::lr_handle_init_*. This commit modifies the function so that
it can be used from the Repo::Impl code and cleans it up.

The only functional change should be that the user agent is now only set
for a remote handle in Repo::Impl.

Commit:
libdnf/repo: Clean up Repo::Impl::lr_handle_init_*

Move duplcate code to common_handle_setup(), further cleanups in
lr_handle_init_remote().

Commit:
libdnf/repo: Move generic remote handle setup to new_remote_handle()

Some of the handle setup in Repo::Impl is generic (not repo metadata
specific) and should be done for a regular librepo handle too.

Also drops setting default values for LRO_CONNECTTIMEOUT and
LRO_LOWSPEEDTIME, as those default values are already set in librepo.

Commit:
libdnf/repo: Simplify logic in lr_handle_perform()

jrohel commented:
* The newly introduced file `package_downloader.hpp` is a public header file. Please, move it to the `include` subdirectory: `include/libdnf/repo/package_downloader.hpp`.
* Typo in the commit message in the commit "Use TempDir in Repo downloading code": "uses th class" -> "uses the class" .

I did some code deduplication in libdnf4 and started the port to libdnf5 a long time ago: https://github.com/rpm-software-management/libdnf/pull/1160. So, I can close my old unfinished PR after merging this. :-)

jrohel commented:
There is a bug in your PR. Callbacks for download packages not working.
Simple test - output of `microdnf download iftop atop`.

Before your PR:
```
Downloading Packages:
[1/2] iftop-0:1.0-0.26.pre4.fc35.x86_64                       100% | 817.7 KiB/s |  51.5 KiB |  00m00s
[2/2] atop-0:2.6.0-7.fc35.x86_64                              100% |   2.0 MiB/s | 171.1 KiB |  00m00s
------------------------------------------------------------------------------------------------------
[2/2] Total                                                   100% | 251.8 KiB/s | 222.6 KiB |  00m01s
```
With your PR:
```
Downloading Packages:
------------------------------------------------------------------------------------------------------
[0/2] Total                              100% [ <=>              ] |   0.0   B/s |   0.0   B |  00m01s
```

lukash commented:
@jrohel thanks, I've fixed the progressbars, moved the header and fixed the typo, I also renamed the `PackageDownloadCallback` class to `PackageDownloadCallbacks` to unify the naming.

jrohel commented:
Thanks. I will merge it.
